src/tqec/sketchup/zx_graph.py

```diff
@@ -279,3 +353,109 @@ def __eq__(self, other: object) -> bool:
         if not isinstance(other, ZXGraph):
             return False
         return cast(bool, nx.utils.graphs_equal(self._graph, other._graph))
+
+    def _find_correlation_subgraphs_dfs(
+        self,
+        parent_corr_node: ZXNode,
+        parent_zx_node_type: NodeType,
+        visited_positions: set[Position3D],
+    ) -> list[set[ZXEdge]]:
+        correlation_subgraphs: list[set[ZXEdge]] = []
+        parent_position = parent_corr_node.position
+        parent_corr_type = parent_corr_node.node_type
+        visited_positions.add(parent_position)
+
+        branched_subgraphs: list[list[set[ZXEdge]]] = []
+        for edge in self.edges_at(parent_position):
+            cur_zx_node = edge.u if edge.v.position == parent_position else edge.v
+            if cur_zx_node.position in visited_positions:
+                continue
+            cur_corr_type = (
+                parent_corr_type.dual() if edge.has_hadamard else parent_corr_type
+            )
+            cur_corr_node = ZXNode(cur_zx_node.position, cur_corr_type)
+            edge_between_cur_parent = ZXEdge(
+                parent_corr_node, cur_corr_node, edge.has_hadamard
+            )
+            branched_subgraphs.append(
+                [
+                    subgraph | {edge_between_cur_parent}
+                    for subgraph in self._find_correlation_subgraphs_dfs(
+                        cur_corr_node, cur_zx_node.node_type, set(visited_positions)
```

Comment:
I think using `copy` would be more explicit here
```suggestion
                        cur_corr_node, cur_zx_node.node_type, copy(visited_positions)
```